Move to jsr363 dependencies

dispiste opened the pull request:
Proposal: https://github.com/geotools/geotools/wiki/Migrate-Units-to-JSR-363

https://osgeo-org.atlassian.net/browse/GEOT-5867

jodygarnett commented:
Thank you, I will continue here for a bit...

jodygarnett commented on modules/library/referencing/pom.xml:
```diff
@@ -214,6 +214,11 @@
+    <dependency>
+      <groupId>si.uom</groupId>
+      <artifactId>systems-common</artifactId>
+      <version>0.7.2</version>
```
we should get this from dependency management

jodygarnett commented:
So I am running into trouble with gt-coverage which makes use of JSCR-275 Measure:

* http://jscience.org/api/javax/measure/Measure.html

With Measure<V,Q extends Quantity>

> This class represents the result of a measurement stated in a known unit.
> There is no constraint upon the measurement value itself: scalars, vectors, or even data sets are valid values as long as an aggregate magnitude can be determined (see Measurable).

Examples provided for:

* DecimalMeasure - a measure whose value is an arbitrary-precision decimal number
* VectorMeasure - a measurement vector of two or more dimensions

I cannot seem to find a similar JSR-363 representation.

dispiste commented:
I believe that JSR363 intended to represent Measure using particular
specializations of the Quantity interface, which can be defined
application-wise.
For instance, the specification (pages 21-22) provides an example of
quantity:

Vector3D<Speed> aircraftSpeed = new Vector3D(12.0, 34.0, -45.5,
METRE_PER_SECOND);

Vector3D is not defined in the specification, and I assume it is just an
example of what can be constructed based on Quantity.

So maybe we need to define Measure and Measurable in Geotools, extending
the Quantity interface, following the implementation provided by JSR275. I
see there is only 4 classes making use of Measure and none of them is
making use of any particular implementation of Measure (such as
DecimalMeasure or VectorMeasure), so I believe that defining Measure and
Measurable would be enough.

Opinions?

César


On 27 March 2018 at 01:36, Jody Garnett <notifications@github.com> wrote:

> So I am running into trouble with gt-coverage which makes use of JSCR-275
> Measure:
>
>    - http://jscience.org/api/javax/measure/Measure.html
>
> With Measure<V,Q extends Quantity>
>
> This class represents the result of a measurement stated in a known unit.
> There is no constraint upon the measurement value itself: scalars,
> vectors, or even data sets are valid values as long as an aggregate
> magnitude can be determined (see Measurable).
>
> Examples provided for:
>
>    - DecimalMeasure - a measure whose value is an arbitrary-precision
>    decimal number
>    - VectorMeasure - a measurement vector of two or more dimensions
>
> I cannot seem to find a similar JSR-363 representation.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/geotools/geotools/pull/1834#issuecomment-376346508>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABfd32omLS3piP1r9MPq1HCMyU9cu3GDks5tiXuHgaJpZM4Sxnp9>
> .
>



-- 
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
   César Martínez Izquierdo
   GIS developer
   -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -
   SCOLAB: http://www.scolab.es
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -

aaime commented:
Hi Jody,
if you cannot find usages of those classes I'm more than happy to get rid
of them :)

Cheers
Andrea

On Tue, Mar 27, 2018 at 1:36 AM, Jody Garnett <notifications@github.com>
wrote:

> So I am running into trouble with gt-coverage which makes use of JSCR-275
> Measure:
>
>    - http://jscience.org/api/javax/measure/Measure.html
>
> With Measure<V,Q extends Quantity>
>
> This class represents the result of a measurement stated in a known unit.
> There is no constraint upon the measurement value itself: scalars,
> vectors, or even data sets are valid values as long as an aggregate
> magnitude can be determined (see Measurable).
>
> Examples provided for:
>
>    - DecimalMeasure - a measure whose value is an arbitrary-precision
>    decimal number
>    - VectorMeasure - a measurement vector of two or more dimensions
>
> I cannot seem to find a similar JSR-363 representation.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/geotools/geotools/pull/1834#issuecomment-376346508>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAT3OYuJsJ7hiIzT_SwxLme3bpAw-pavks5tiXuFgaJpZM4Sxnp9>
> .
>



-- 

Regards,

Andrea Aime

==
GeoServer Professional Services from the experts! Visit http://goo.gl/it488V
for more information.
==

Ing. Andrea Aime
@geowolf
Technical Lead

GeoSolutions S.A.S.
Via di Montramito 3/A
55054  Massarosa (LU)
phone: +39 0584 962313
fax: +39 0584 1660272
mob: +39  339 8844549

http://www.geo-solutions.it
http://twitter.com/geosolutions_it

AVVERTENZE AI SENSI DEL D.Lgs. 196/2003

Le informazioni contenute in questo messaggio di posta elettronica e/o
nel/i file/s allegato/i sono da considerarsi strettamente riservate. Il
loro utilizzo è consentito esclusivamente al destinatario del messaggio,
per le finalità indicate nel messaggio stesso. Qualora riceviate questo
messaggio senza esserne il destinatario, Vi preghiamo cortesemente di
darcene notizia via e-mail e di procedere alla distruzione del messaggio
stesso, cancellandolo dal Vostro sistema. Conservare il messaggio stesso,
divulgarlo anche in parte, distribuirlo ad altri soggetti, copiarlo, od
utilizzarlo per finalità diverse, costituisce comportamento contrario ai
principi dettati dal D.Lgs. 196/2003.

The information in this message and/or attachments, is intended solely for
the attention and use of the named addressee(s) and may be confidential or
proprietary in nature or covered by the provisions of privacy act
(Legislative Decree June, 30 2003, no.196 - Italy's New Data Protection
Code).Any use not in accord with its purpose, any disclosure, reproduction,
copying, distribution, or either dissemination, either whole or partial, is
strictly forbidden except previous formal approval of the named
addressee(s). If you are not the intended recipient, please contact
immediately the sender by telephone, fax or e-mail and delete the
information in this message that has been received in error. The sender
does not give any warranty or accept liability as the content, accuracy or
completeness of sent messages and accepts no responsibility  for changes
made after they were sent or for other risks which arise as a result of
e-mail transmission, viruses, etc.

jodygarnett commented:
@tbarsballe and myself did a desk review:

a) Some similarity of purpose with org.geotools.measure.Measure - but it is a different API contract. It is not worth harmonizing these two interfaces.

b) setting up a org.geotools.coverage.io.range.Measure interface seems to be the way to go, this interface is only used here.

I agree @aaime that we can only find this used in test cases, I will try deprecating all of these for removal (If we do not find any use in the GWC and GeoServer code bases)

jodygarnett commented:
@aaime you are correct those classes were not used.

This leaves me with the codebase compiling, and the following errors in gt-referencing:

* ScriptTest.testStereographic:168->runScript:55 » InvalidParameterValue Value -...
* GeostationarySatelliteTest.testCircumscribeFullDisk_Spheroidal:106 » Projection
* GeostationarySatelliteTest.testInscribeFullDiskEstimate_Ellipsoidal:302 » Projection
* GeostationarySatelliteTest.testInscribeFullDiskEstimate_Spheroidal:250 » Projection
* GeostationarySatelliteTest.testCircumscribeFullDisk_Ellipsoidal:178 » Projection

All of them are range errors of some sort:

```
org.opengis.parameter.InvalidParameterValueException: Value -90 is out of range [-90..90].
```

```
WARNING: Possible use of "GEOS" projection outside its valid area.
Longitude 216°19.2'W is out of range (±180°).
```

So the expectation is that the previous implementation handled angles more elegently.

jodygarnett commented:
@dispiste we are down to a couple test failures, looks to be a regression in our SexagesimalConverter (about the only custom converter we have).

On master GeostationarySatelliteTest:

* [x] testCircumscribeFullDisk_Ellipsoidal()
* [x] testCircumscribeFullDisk_Spheroidal()
* [x] testEllipsoidalWKTParameters()
* [x] testInscribeFullDiskEstimate_Ellipsoidal()
* [x] testInscribeFullDiskEstimate_Spheroidal()

Each test:

* Calls SexagesimalConverter.Invert.convert(10000.0) --> 1.0
* Calls SexagesimalConverter.Invert.convert(1.0) --> 2.777777777777778E-4
* producing the following log (From MapProjection.verifyGeographicRanges) as above

  > WARNING: Possible use of "GEOS" projection outside its valid area.
  > Longitude 216°19.2'W is out of range (±180°).

On master ScriptTest:

* [x] ScriptTest.testStereographic calls
  
  * Calls SexagesimalConverter.Invert.convert(10000.0) --> 1.0
  * Calls SexagesimalConverter.Invert.convert(1.0) --> 2.777777777777778E-4
  * No warning

dispiste commented:
@jodygarnett I'll come back today to this task and try to fix the remaining issues.

jodygarnett commented:
Current status, GeostationarySatelliteTest are failing, cannot seem to trace it down yet. Comparing debugger runs between branches is a pain.

jodygarnett commented on modules/library/referencing/src/main/java/org/geotools/measure/Units.java:
```diff
@@ -97,4 +100,27 @@ private Units() {
+                MultiplyConverter multiplyConverter = (MultiplyConverter) converter;
+                double factor = multiplyConverter.getFactor();
+                if (factor == 0.017453292519943295 || factor == 0.0174532925199433
+                        || factor == 0.01745329251994328) {
```
Better implemented with:

```
    static final double RADIAN_TO_DEGREE_RATIO = Math.PI/180.0; // 0.017453292519943295
    
...
    if (Math.abs(RADIAN_TO_DEGREE_RATIO - factor) < 1.0e-17) {
```

jodygarnett commented:
Comparing [GeostationarySatelliteTest Line 76|GeostationarySatelliteTest github]:
```
ellipsoidalGeosCRS = CRS.parseWKT(ellipsoidalGeosWKT);
```

Parsing:
```
    public final static String ellipsoidalGeosWKT =
            "PROJCS[\"Geostationary_Satellite\"," +
                    "  GEOGCS[\"WGS 84\"," +
                    "    DATUM[\"WGS_1984\"," +
                    "      SPHEROID[\"WGS84\",6378137,298.257223563]]," +
                    "    PRIMEM[\"Greenwich\",0],UNIT[\"degree\",0.01745329251994328]]," +
                    "    PROJECTION[\"Geostationary_Satellite\"]," +
                    "    PARAMETER[\"central_meridian\", -135]," +
                    "    PARAMETER[\"satellite_height\",35785831.0]," +
                    "    PARAMETER[\"false_easting\",0]," +
                    "    PARAMETER[\"false_northing\",0]," +
                    "    UNIT[\"meter\", 1]]";
```

Master produces UNIT["degree", 0.017453292519943295]: 

```
PROJCS["Geostationary_Satellite", 
  GEOGCS["WGS 84", 
    DATUM["WGS_1984", 
      SPHEROID["WGS84", 6378137.0, 298.257223563]], 
    PRIMEM["Greenwich", 0.0], 
    UNIT["degree", 0.017453292519943295], 
    AXIS["Longitude", EAST], 
    AXIS["Latitude", NORTH]], 
  PROJECTION["GEOS"], 
  PARAMETER["central_meridian", -135.0], 
  PARAMETER["satellite_height", 35785831.0], 
  PARAMETER["false_easting", 0.0], 
  PARAMETER["false_northing", 0.0], 
  UNIT["m", 1.0], 
  AXIS["x", EAST], 
  AXIS["y", NORTH]]
```

The branch produces UNIT["rad*0.01745329251994328", 0.01745329251994328]:

```
PROJCS["Geostationary_Satellite", 
  GEOGCS["WGS 84", 
    DATUM["WGS_1984", 
      SPHEROID["WGS84", 6378137.0, 298.257223563]], 
    PRIMEM["Greenwich", 0.0], 
    UNIT["rad*0.01745329251994328", 0.01745329251994328], 
    AXIS["Lon", EAST], 
    AXIS["Lat", NORTH]], 
  PROJECTION["GEOS"], 
  PARAMETER["central_meridian", -135.0], 
  PARAMETER["satellite_height", 35785831.0], 
  PARAMETER["false_easting", 0.0], 
  PARAMETER["false_northing", 0.0], 
  UNIT["m", 1.0], 
  AXIS["x", EAST], 
  AXIS["y", NORTH]]
```

bencaradocdavies commented on modules/library/referencing/src/main/java/org/geotools/measure/Units.java:
```diff
@@ -130,8 +132,7 @@ public static final boolean isDegreeAngle(Unit<?> unit) {
                 double factor = multiplyConverter.getFactor();
-                if (factor == 0.017453292519943295 || factor == 0.0174532925199433
-                        || factor == 0.01745329251994328) {
+                if (Math.abs(RADIAN_TO_DEGREE_RATIO - factor) < 1.0e-17) {
```
@jodygarnett I would prefer a tolerance as a static final constant. This also looks too small too me. Subtractive cancellation can easily lose several orders of magnitude of precision. If I compare the three constants removed with the tolerance, you can already see that the last will already differ by more than the the tolerance:
```
0.017453292519943295
0.017453292519943300
0.017453292519943280
0.000000000000000010
```
How about 1e-15?

jodygarnett replied:
Thanks Ben, relaxing this tolerance has referencing passing (especially now that @dispiste uses the test while parsing)

jodygarnett commented:
Next failure is in gt-epsg-hsql EsriLookupTest:
* Could not find code for 2264
* Could not find code for 26802

Looking at the WKT they differ on:

* expected: UNIT["Foot_US",0.3048006096012192]
  
  actual: UNIT["m*0.3048006096012192", 0.3048006096012192]

* expected: UNIT["Degree",0.0174532925199433]
  
  actual: UNIT["degree", 0.017453292519943295]

jodygarnett commented:
Okay the rest of this seems to just be work, inserting auto-correct unit matching to SE encoding/formatting, geotiff reading/writing, etc...

jodygarnett commented:
BUILD SUCCEEDS locally, now waiting for travis.

Changes made while getting tests to pass:

* Removed use of UCUM: a) UCUM would not load due to version differences (missing drag class) b) everything was already in USCustomary
* Made a utility method to autocorrect units that were close to a built-in unit
* Made heavy use of the SimpleFormat.getInstance(), to label common units (Kelvin) and alias others 

I would like to ask for a review while we update GWC and GeoServer codebase.

jodygarnett commented on modules/plugin/coverage-multidim/netcdf/src/main/java/org/geotools/imageio/netcdf/VariableAdapter.java:
```diff
@@ -883,7 +884,7 @@ private void initRange() {
                 unit = SimpleUnitFormat.getInstance().parse(unitString);
                
-            } catch (IllegalArgumentException iae) {
+            } catch (ParserException parseException) {
```
@aaime NetCDF failures came down to this one change, the exception changed. It is thrown when parsing unitsString "1" for example.

jodygarnett commented on modules/library/coverage/src/main/java/org/geotools/coverage/grid/io/imageio/geotiff/CRS2GeoTiffMetadataAdapter.java:
```diff
@@ -978,6 +979,8 @@ private void parseUnit(Unit<?> unit, int model,GeoTiffIIOMetadataEncoder metadat
                 default:
                     throw new IllegalStateException("Unable to map model "+model+" for this unit");
 	        }
+	        unit = Units.autoCorrect(unit); // auto-correct DEGREE_ANGLE and FOOT_SURVEY
```
@aaime fixing GeoTIFF came down to "auto-correcting" units when reading.

jodygarnett commented on modules/library/referencing/src/main/java/org/geotools/measure/Units.java:
```diff
@@ -115,24 +115,70 @@ public static void registerCustomUnits(SimpleUnitFormat format) {
+     * @param unit 
+     * @return Unit, possibly auto-corrected based on multiplication factor match
+     */
+    public static Unit<?> autoCorrect(Unit<?> unit) {
```
This is they key method, identifying common multiplication factors, and providing the built-in unit as a replacement.

jodygarnett commented on modules/library/referencing/src/main/java/org/geotools/measure/Units.java:
```diff
@@ -96,9 +96,24 @@ private Units() {
+     * 
+     * @param format The UnitFormat in which the labels and aliases must be registered.
+     */
+    public static void registerCustomUnits(SimpleUnitFormat format) {
```
This is the other key fix, provided by @dispiste, that is used to configure our built-in SimpleUnitFormat instances with common labels and alias.  Subsequent lists add to this list (for pixel and Kelvin) but do not change the approach.

dispiste commented:
Thanks @jodygarnett and all the others for completing the pull request.

I am in general satisfied with the result, but I still see some
potential regressions in WKT parsing or CRS.lookupEpsgCode in cases
not covered by unit tests.
In particular, I've seen that the old library was able to identify
equivalent units in the equals method, as far as their equivalent
factor was the same.
For instance, US survey foot could be defined using a fraction like this:

Unit<Length> FOOT_SURVEY = addUnit(METER.multiply(1200).divide(3937),
"US Survey foot", "ft_survey_us");
which internally uses a RationalConverter to define the unit.

But it could also be defined as:
Unit<Length> FOOT_SURVEY = addUnit(METER.multiply(0.3048006096012192),
"US Survey foot", "ft_survey_us");
which internally uses a MultiplyConverter.

The old implementation was able to identify these definitions as
equivalent in the equals method [1], while the new one considers
definitions are equivalent if they use the same converter class (so a
RationalConverter will never be considered equal to a
MultiplyConverter, see [2] and [3]).

We are currently catching these cases for US survey feet and degrees,
but some other cases may need a similar correction and are not easy to
spot until downstream products migrate to gt20 and start reporting
them. This would require adding new exceptions to Units.autocorrect
(or fixing the equals method in the uom-se implementation).


[1] https://github.com/tarelli/jscience/blob/28bf87f3e98dde9524e877bade78d5a32728f65e/src/javax/measure/converter/UnitConverter.java#L84
[2] https://github.com/unitsofmeasurement/uom-se/blob/1bdaeb20e02d6f521735a0b9697988da676d6a17/src/main/java/tec/uom/se/function/RationalConverter.java#L219
[3] https://github.com/unitsofmeasurement/uom-se/blob/1bdaeb20e02d6f521735a0b9697988da676d6a17/src/main/java/tec/uom/se/function/MultiplyConverter.java#L116

On 4 April 2018 at 18:58, Jody Garnett <notifications@github.com> wrote:
> Next failure is in gt-epsg-hsql EsriLookupTest:
>
> Could not find code for 2264
> Could not find code for 26802
>
> Looking at the WKT they differ on:
>
> expected: UNIT["Foot_US",0.3048006096012192]
>
> actual: UNIT["m*0.3048006096012192", 0.3048006096012192]
>
> expected: UNIT["Degree",0.0174532925199433]
>
> actual: UNIT["degree", 0.017453292519943295]
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.



-- 
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
   César Martínez Izquierdo
   GIS developer
   -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -
   SCOLAB: http://www.scolab.es
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -

jodygarnett commented:
With respect to depending on SI, NonSI and USCustomary.

This locks us into one implementation of units.

If we did make a series of constants for units as you describe, could we make use of unit parsing rather than a direct dependency? This way we would allow applications to choose their own implementation.

dispiste commented on modules/library/metadata/src/main/java/org/geotools/util/NumberRange.java:
```diff
@@ -442,7 +442,8 @@ public boolean contains(final Number value) throws IllegalArgumentException {
      */
     @Override
-    public boolean contains(@SuppressWarnings("rawtypes") Comparable value) throws IllegalArgumentException {
+    public boolean contains(@SuppressWarnings("rawtypes") Comparable<?> value)
```
The previous commit in the branch contained:
```java
    public boolean contains(@SuppressWarnings("rawtypes") Comparable value) throws IllegalArgumentException {
```
Master contained:
```java
public boolean contains(Comparable<?> value) throws IllegalArgumentException {
```
which was not considered the same type at compile time, so it was breaking Geoserver compilation.
Therefore, I restored a signature more similar to the one present in master.

dispiste commented:
- Regarding SI, NonSI and USCustomary: This would be an interesting option. In that case, would you mark the dependency on the units library as optional in Geotools poms? Can this create any problem to Geotools users?

- Regarding the problem with Converters. The equals method is used internally by the library to decide certain things (for instance in the "unitFor" and "nameFor" methods in SimpleUnitFormat), so implementing our Units.equals would not solve the issue. I think we can just hope that no other units stop being parsed, or write more unit tests which try to parse additional units to be sure. And write additional cases in Units.autocorrect if needed.

jodygarnett commented:
Thanks @dispiste I am sensitive to this PR being open long enough to be an integration risk. It may be preferable to merge this sooner (the api change) and then optimize/improve later (better/deeper matching)

aaime commented:
@jodygarnett I'm sensitive to another "filter upgrade" situation, many years later we still have a lot of deprecated stuff that should have been removed later. When you merge make sure you have something that can be considered"final" :-)

aaime commented:
Lots of changes seem to involve unused imports. Guess it's the result of a "sed" over the files. Not a problem per se, just an observation (that the pull request is actually less invasive than then number of files changed may make it look like).

aaime commented:
There is also a lot of reformats and import changes... I'm ignoring them since we are going to reformat everything anyways, but for the future, try to minimize those changes.

aaime commented:
Too big, cannot finish it today. I'll try to resume tomorrow (already set various comments in the review, but it's not complete)

dispiste commented:
Hi @jodygarnett @aaime

I've committed some additiona fixes, but I there are some fundamental issues (in addition to the minor corrections pointed by @aaime in the review) which IMHO should be solved before merging (or at least they should be considered when approving the migration). Unfortunately I will not be able to invest more time in this JSR363 migration (I've already spent quite a lot), so I've taken some time to explain the issues and I've also outlined 2 different solutions to avoid them (one of them is implemented in a new subbranch, see details below).

The root of the problem: the old JSR275 library was able to identify equivalent unit definitions in the equals method, while the new one is not able to do so when the units to compare are defined using different converter classes (e.g. RationalConverter vs MultiplyConverter).

This different behaviour has an impact at least in the following Geotools use cases, potentially causing regresions:

1.  method CRS.lookupEpsgCode(crs, false): when trying to derive an EPSG code from a CRS created from a non-EPSG WKT string, because the crs units are not considered equivalent units to the ones defined in the EPSG database and the replacement fails

2. method CRS.toWKT(): when trying to format a unit. In the resulting WKT string, units that have not been recognised will be assigned a name based in the base unit and the conversion factor. Example using JSR-363 implementation: UNIT["m*0.3048006096012192", 0.3048006096012192], instead of UNIT["foot_survey_us", 0.3048006096012192] using the old JSR-275 library.

3. when parsing/writing units for Geotiff reading/writing (see CRS2GeoTiffMetadataAdapter.parseLinearUnit and parseUnit methods)

4. in any place where the getConverterTo()/getConverterToAny() method is used: it will produce a non-Identity converter if the fromUnit and toUnit aren't considered equivalent by the new library. This creates precision errors as illustrated in the following examples:
```java
// old lib: returns 1.0
NonSI.DEGREE_ANGLE.getConverterTo(SI.RADIAN.times(0.0174532925199433)).convert(1.0)
// new lib: returns 0.9999999999999998
NonSI.DEGREE_ANGLE.getConverterTo(SI.RADIAN.multiply(0.0174532925199433)).convert(1.0)
// old lib: returns 23.0
NonSI.DEGREE_ANGLE.getConverterTo(SI.RADIAN.times(0.0174532925199433)).convert(23.0)
// new lib: returns 22.999999999999993
NonSI.DEGREE_ANGLE.getConverterTo(SI.RADIAN.multiply(0.0174532925199433)).convert(23.0)
```

The behaviour of the new implementation can't be considered incorrect (it is more precise), but Geotools is not expecting this behaviour and it will produce different results and even errors in some situations.

In order to mitigate these issues, Units.autocorrect method was introduced, which tries to replace the provided unit by another well-known unit. This is done case by case (currently only for 2 units), so it mitigates the problem but does not prevent it to happen for other units that have not been condidered in autocorrect.

We can envisage a full solution to these issues using one of the following approaches:

1.  We can't inject the old behaviour in the JSR363 implementation, but we can implement a Units.equals(Unit<?> unit1, Unit<?> unit2) method which mimics the JSR-275 behaviour and use it in key points of Geotools. In addition, we will also override the forName method in EPSGFormat and ESRIFormat classes to use equivalent names when available. Finally, we will provide an Units.getConverterToAny method which provides the Identity converter if units are considered equivalent. Note that this approach might still fail if we have failed to identify some of the key points in which Units.equals should be used. I've created a subbranch implementing this solution, which compiles and passes tests (even disabling the autocorrect method), but I find the changes quite intrusive. On the TODO list: replace all the usages of getConverterTo()/getConverterToAny()  by the Units.getConverterToAny() method. The branch is here: units_jsr363_equalsapproach

2. We can try to get one of the units defined by the new implementation whenever we need to instantiate a unit. We would parse the unit name to get a unit instance, then we would compare the conversion factor of the instatiated unit with the provided provided factor. In this way, we will not need to later check for units equivalences and we would avoid all the mentioned issues. The will produce a side effect when we parse a unit such  UNIT["degree", 0.0174532925199433], since it will inmediately be converted on the reference degree unit and we will loose some information (the real conversion factor defined in the WKT). Anyway, this is partially happening alredy in the JSR275 implementation and also in the S1 solution, so I think it is acceptable. I find this approach cleaner and simpler and I'm trying to implement it on a new subbranch.

Finally, note that some small differences will remain compared with the JSR275 implementation:

- Units may have a different label, which has some effects in formatting. For instance, old library defined "foot_survey_us" while the new one defines "ft_survey_us". It doesn't seem a big problem, since none of these labels is the official one defined by the EPSG database: "US Survey foot". In general, names seem to be ignored when parsing units. In any case, we can override labels in Units.registerCustomUnits (system-wide), in the EPSGFormat constructor (epsg-wide) or in the ESRIFormat constructor (esri-wide)

I hope this information is useful to take a decision for next steps.

dispiste commented:
Hi again, I've finally done a last try and implemented an approach similar to the second solution outlined above. Basically, it consists on replacing the existing autocorrect method by a more generic one which mimics behaviour of the old JSR275 library. I also tried to use autocorrect only where necessary (for units instantiation), so I removed it from some places. I am quite satisfied with the result, so I've committed it to this branch. It compiles and passes tests.

I believe that now only @aaime comments needs to be tackled to get the PR ready for merging.

I've written some documentation regarding units usage, which might be added to the proposal:


## JSR363 Units HOWTO
### Getting Unit instances
If you know the unit to use **at compile time**, use one of the Unit instances defined as static variables in `org.geotools.measure.Units`, `si.uom.SI`, `si.uom.NonSI` or `systems.uom.common.USCustomary`.

If you need to define new Units **at runtime**, it is important to **immediately try to convert the new unit to one of the existing instances using `Units.autocorrect` method**.  Autocorrect applies some tolerance to locate an equivalent Unit. Skipping autocorrect will produce unexpected results and errors due to small differences in units definition. Examples:
```java
// the result should be NonSI.DEGREE_ANGLE:
Unit<?> deg = Units.autoCorrect(SI.RADIAN.multiply(0.0174532925199433));
```
```java
Unit<?> halfMetre = SI.METRE.divide(2);
// the result should be SI.METRE
Unit<?> stupidUnit = Units.autocorrect(halfMetre.multiply(4).divide(2));
```
```java
public <T extends Quantity<T>> Unit<T> deriveUnit(Unit<T>  baseUnit, double factor) {
    return Units.autocorrect(baseUnit.multiply(factor);)
}
```
Use a specific `Quantity` whenever possible to get type-safety checks at compile time:
```java
Unit<Length> halfMetre = SI.METRE.divide(2);
Unit<Length> stupidUnit = Units.autocorrect(halfMetre.multiply(4).divide(2));
```

### Formatting units

Use `org.geotools.measure.Units.toName()` to get the unit name (or unit label if name is not defined). Use `org.geotools.measure.Units.getDefaultFormat().format()` to get the unit label (ignoring the name). Examples:
```java
Unit<?> unit = ...
System.out.println(Units.toName(unit)):
```

```java
// prints "Litre"
System.out.println(Units.toName(SI.LITRE))
// prints "l"
System.out.println(Units.getDefaultFormat().format(SI.LITRE))
// Most units don't define a name, so it does not make a difference
// prints "m"
System.out.println(Units.toName(SI.METRE))
// prints "m"
System.out.println(Units.getDefaultFormat().format(SI.METRE))
```

There are specific formatters for EPSG and ESRI dialects (although they are not public classes, they are used by WKT `Formatter` class), which override the default formatting (for instance EPSG formatter prints "Metre" instead of "m").
Labels can be overridden  in Units.registerCustomUnits (system-wide), in the EPSGFormat.epsgLabelsAndAliases mehtod (epsg-wide) or in the ESRIFormat.esriLabelsAndAliases constructor (esri-wide)

### Converting units

If the unit Quantity type is known, use the type-safe *getConverterTo()* method:
```java
Unit<Angle> unit = ...
UnitConverter converter = unit.getConverterTo(SI.RADIAN);
double convertedQuantity = converter.convert(3.1415);
```

If the Quantity type is undefined, use the convenience method `org.geotools.measure.Units.getConverterToAny()`. Note that this method throws an IllegalArgumentException if units can't be converted:
```java
Unit<?> unit = ...
UnitConverter converter = Units.getConverterToAny(unit, SI.RADIAN);
double convertedQuantity = converter.convert(3.1415);

aaime commented:
On Tue, Apr 17, 2018 at 12:04 PM, Cesar Martinez Izquierdo <
notifications@github.com> wrote:

> Hi @jodygarnett <https://github.com/jodygarnett> @aaime
> <https://github.com/aaime>
>
> I've committed some additiona fixes, but I there are some fundamental
> issues (in addition to the minor corrections pointed by @aaime
> <https://github.com/aaime> in the review) which IMHO should be solved
> before merging (or at least they should be considered when approving the
> migration). Unfortunately I will not be able to invest more time in this
> JSR363 migration (I've already spent quite a lot), so I've taken some time
> to explain the issues and I've also outlined 2 different solutions to avoid
> them (one of them is implemented in a new subbranch, see details below).
>
> The root of the problem: the old JSR275 library was able to identify
> equivalent unit definitions in the equals method, while the new one is not
> able to do so when the units to compare are defined using different
> converter classes (e.g. RationalConverter vs MultiplyConverter).
>
> This different behaviour has an impact at least in the following Geotools
> use cases, potentially causing regresions:
>
>    1.
>
>    method CRS.lookupEpsgCode(crs, false): when trying to derive an EPSG
>    code from a CRS created from a non-EPSG WKT string, because the crs units
>    are not considered equivalent units to the ones defined in the EPSG
>    database and the replacement fails
>
> Yep, this is serious, and it's going to make a sizeable amount of work I
did in this area regress, but maybe not a stopper


>
>    1. method CRS.toWKT(): when trying to format a unit. In the resulting
>    WKT string, units that have not been recognised will be assigned a name
>    based in the base unit and the conversion factor. Example using JSR-363
>    implementation: UNIT["m*0.3048006096012192", 0.3048006096012192], instead
>    of UNIT["foot_survey_us", 0.3048006096012192] using the old JSR-275 library.
>
> This seem like a stopper, but I'm not sure, afraid it will cause issues
with shapefile/world files we generate (or else, with other systems ability
to deal with it).


>    1.
>
>    in any place where the getConverterTo()/getConverterToAny() method is
>    used: it will produce a non-Identity converter if the fromUnit and toUnit
>    aren't considered equivalent by the new library. This creates precision
>    errors as illustrated in the following examples:
>
> // old lib: returns 1.0NonSI.DEGREE_ANGLE.getConverterTo(SI.RADIAN.times(0.0174532925199433)).convert(1.0)// new lib: returns 0.9999999999999998NonSI.DEGREE_ANGLE.getConverterTo(SI.RADIAN.multiply(0.0174532925199433)).convert(1.0)// old lib: returns 23.0NonSI.DEGREE_ANGLE.getConverterTo(SI.RADIAN.times(0.0174532925199433)).convert(23.0)// new lib: returns 22.999999999999993NonSI.DEGREE_ANGLE.getConverterTo(SI.RADIAN.multiply(0.0174532925199433)).convert(23.0)
>
> The behaviour of the new implementation can't be considered incorrect (it
> is more precise), but Geotools is not expecting this behaviour and it will
> produce different results and even errors in some situations.
>

Unsure about this one (if it's a stopper or not)


> In order to mitigate these issues, Units.autocorrect method was
> introduced, which tries to replace the provided unit by another well-known
> unit. This is done case by case (currently only for 2 units), so it
> mitigates the problem but does not prevent it to happen for other units
> that have not been condidered in autocorrect.
>
> We can envisage a full solution to these issues using one of the following
> approaches:
>
>    1.
>
>    We can't inject the old behaviour in the JSR363 implementation, but we
>    can implement a Units.equals(Unit unit1, Unit unit2) method which mimics
>    the JSR-275 behaviour and use it in key points of Geotools. In addition, we
>    will also override the forName method in EPSGFormat and ESRIFormat classes
>    to use equivalent names when available. Finally, we will provide an
>    Units.getConverterToAny method which provides the Identity converter if
>    units are considered equivalent. Note that this approach might still fail
>    if we have failed to identify some of the key points in which Units.equals
>    should be used. I've created a subbranch implementing this solution, which
>    compiles and passes tests (even disabling the autocorrect method), but I
>    find the changes quite intrusive. On the TODO list: replace all the usages
>    of getConverterTo()/getConverterToAny() by the
>    Units.getConverterToAny() method. The branch is here:
>    units_jsr363_equalsapproach
>
> Interesting, I'll try to have a look

>
>    1. We can try to get one of the units defined by the new
>    implementation whenever we need to instantiate a unit. We would parse the
>    unit name to get a unit instance, then we would compare the conversion
>    factor of the instatiated unit with the provided provided factor. In this
>    way, we will not need to later check for units equivalences and we would
>    avoid all the mentioned issues. The will produce a side effect when we
>    parse a unit such UNIT["degree", 0.0174532925199433], since it will
>    inmediately be converted on the reference degree unit and we will loose
>    some information (the real conversion factor defined in the WKT). Anyway,
>    this is partially happening alredy in the JSR275 implementation and also in
>    the S1 solution, so I think it is acceptable. I find this approach cleaner
>    and simpler and I'm trying to implement it on a new subbranch.
>
> Yep, agreed.


> Finally, note that some small differences will remain compared with the
> JSR275 implementation:
>
>    - Units may have a different label, which has some effects in
>    formatting. For instance, old library defined "foot_survey_us" while the
>    new one defines "ft_survey_us". It doesn't seem a big problem, since none
>    of these labels is the official one defined by the EPSG database: "US
>    Survey foot". In general, names seem to be ignored when parsing units. In
>    any case, we can override labels in Units.registerCustomUnits
>    (system-wide), in the EPSGFormat constructor (epsg-wide) or in the
>    ESRIFormat constructor (esri-wide)
>
> Yeah, I guess it would be useful.

I am worried about the current state, as this is crossing with the
"wholesale reformat" proposal, if this does not get merged
before the reformat goes in, you'll get a landslide of conflicts.
Thankfully the reformat proposal is a "method and tool" that requires very
few manual changes, the reformat will be done on the fly
when we are ready by a tool... but I'd rather not wait for months before
applying it (I'm happy to wait a few weeks more if needed though).

Cheers
Andrea

==

GeoServer Professional Services from the experts! Visit http://goo.gl/it488V
for more information.
==

Ing. Andrea Aime
@geowolf
Technical Lead

GeoSolutions S.A.S.
Via di Montramito 3/A
55054  Massarosa (LU)
phone: +39 0584 962313
fax: +39 0584 1660272
mob: +39  339 8844549

http://www.geo-solutions.it
http://twitter.com/geosolutions_it

AVVERTENZE AI SENSI DEL D.Lgs. 196/2003

Le informazioni contenute in questo messaggio di posta elettronica e/o
nel/i file/s allegato/i sono da considerarsi strettamente riservate. Il
loro utilizzo è consentito esclusivamente al destinatario del messaggio,
per le finalità indicate nel messaggio stesso. Qualora riceviate questo
messaggio senza esserne il destinatario, Vi preghiamo cortesemente di
darcene notizia via e-mail e di procedere alla distruzione del messaggio
stesso, cancellandolo dal Vostro sistema. Conservare il messaggio stesso,
divulgarlo anche in parte, distribuirlo ad altri soggetti, copiarlo, od
utilizzarlo per finalità diverse, costituisce comportamento contrario ai
principi dettati dal D.Lgs. 196/2003.

The information in this message and/or attachments, is intended solely for
the attention and use of the named addressee(s) and may be confidential or
proprietary in nature or covered by the provisions of privacy act
(Legislative Decree June, 30 2003, no.196 - Italy's New Data Protection
Code).Any use not in accord with its purpose, any disclosure, reproduction,
copying, distribution, or either dissemination, either whole or partial, is
strictly forbidden except previous formal approval of the named
addressee(s). If you are not the intended recipient, please contact
immediately the sender by telephone, fax or e-mail and delete the
information in this message that has been received in error. The sender
does not give any warranty or accept liability as the content, accuracy or
completeness of sent messages and accepts no responsibility  for changes
made after they were sent or for other risks which arise as a result of
e-mail transmission, viruses, etc.

jodygarnett commented on modules/library/coverage/src/main/java/org/geotools/coverage/grid/io/imageio/geotiff/CRS2GeoTiffMetadataAdapter.java:
```diff
@@ -970,13 +979,15 @@ private void parseUnit(Unit<?> unit, int model,GeoTiffIIOMetadataEncoder metadat
                     throw new IllegalStateException("Unable to map model "+model+" for this unit");
 	        }
-	        if(unit.equals(SI.METER)){
+	        unit = Units.autoCorrect(unit); // auto-correct DEGREE_ANGLE and FOOT_SURVEY
```
You are correct, so far it has just added it when running into accuracy differences that prevented "matching" units from being recognized as such.

Perhaps Units.toCanonicalUnit(Unit) would be more appropriate.

jodygarnett commented on modules/extension/xsd/xsd-sld/src/test/java/org/geotools/se/v1_1/RountTripLocalUomTest.java:
```diff
@@ -14,7 +15,8 @@
     public void testRoundTripPoint() throws Exception {
         PointSymbolizer sym = (PointSymbolizer) parse("example-pointsymbolizer-local-uom.xml");
-        assertEquals(NonSI.PIXEL, sym.getUnitOfMeasure());
+        assertEquals("pixel", Units.toName( sym.getUnitOfMeasure() ));
```
Strangely we did find a PIXEL constant, as a unit of information measuring 4 bytes :)

jodygarnett commented on modules/library/coverage/src/main/java/org/geotools/coverage/grid/io/imageio/geotiff/CRS2GeoTiffMetadataAdapter.java:
```diff
@@ -996,19 +1007,20 @@ private void parseUnit(Unit<?> unit, int model,GeoTiffIIOMetadataEncoder metadat
-                            }
+                        if (!AbstractUnit.ONE.equals(unit)) {
+                            // or SI.RADIAN
+                            converter = ((Unit<Angle>) unit).getConverterTo(NonSI.SECOND_ANGLE);
```
Base was compatible with radian, was not necessarirly radian. Still I will run test cases again and see if I can make the code more clear

jodygarnett commented:
Thanks for they key update @dispiste, I too am quite happy with the "second approach" result.

dispiste commented:
Hi, I've tackled the requested changes. I've ignored code formatting concerns (they are the result of a lost battle against configuration of Eclipse automatic format), since I understand that a global reformatting is coming soon.

Maybe commit https://github.com/geotools/geotools/pull/1834/commits/7aba0046209d580a7f90b59d157f4d80a260a855 needs to be also reviewed, then we are ready to merge IMHO.

jodygarnett commented:
I have written up your notes and the change proposal as as upgrade documentation.

This is now ready to merge in conjunction with https://github.com/geoserver/geoserver/pull/2828

jodygarnett commented:
Okay so geoserver reports a netCDF regression:


```
Failed tests:   testRapNativeGribRotatedPole(org.geoserver.wcs2_0.WCSNetCDFTest): expected:<[K]> but was:<[kelvin]>
  testNetcdfRotatedPole(org.geoserver.wcs2_0.WCSNetCDFTest): expected:<[K]> but was:<[kelvin]>
```

Difference between using a symbol ``K`` and a name `kelvin`? Or do we need to make another format / configuration for use with netCDF.

jodygarnett commented on modules/plugin/coverage-multidim/netcdf/src/main/java/org/geotools/imageio/netcdf/VariableAdapter.java:
```diff
@@ -880,8 +882,9 @@ private void initRange() {
                 }
-                unit = Unit.valueOf(unitString);
-            } catch (IllegalArgumentException iae) {
+                unit = SimpleUnitFormat.getInstance().parse(unitString);
```
May need to use Units.parseUnit(unitStirng) to access configured units?

dispiste replied:
In Units [1], we define "kelvin" as the label for Kelvin unit. You might want to change it to "K".

Replacing ``` SimpleUnitFormat.getInstance().parse(unitString);``` by ```Units.parseUnit()``` is also a good idea, since it ensures that our units are registered on SimpleUnitFormat (Geotools units are registered during static code execution of Units class).

[1] https://github.com/dispiste/geotools/blob/0b824095825e9287f7e2a16a983a1aa984d29fc2/modules/library/referencing/src/main/java/org/geotools/measure/Units.java#L136

dispiste commented:
JSR363 API provides some methods in Unit interface:
- getName(): it should get the name
- getSymbol(): it should get the symbol

However, the implementation returns null for  getName() and/or  getSymbol() for most units. Some examples:
SI.METRE.getName() returns null
SI.METRE.getSymbol() returns "m"
SI.LITRE.getName() returns "Litre"
SI.LITRE.getSymbol() returns null
NonSI.DEGREE_ANGLE.getName() returns "Degree Angle"
NonSI.DEGREE_ANGLE.getSymbol() returns null
USCustomary.FOOT_SURVEY.getName() returns "US Survey foot"
USCustomary.FOOT_SURVEY.getSymbol() returns null

The toString() method relies on  SimpleUnitFormat.getInstance().format() method.

The SimpleUnitFormat.getInstance().format() method is based on the labels that have been registered using the label() method. If no labels have been registered, format uses getSymbol() for `BaseUnit`s and `AlternateUnit`s. For `TransformedUnit`s, a name is composed based on the BaseUnit format and the conversion factor.

Therefore, from a practical point of view, I would say we only have format() method, which we can configure in the best way we decide, but that it will fallback to getSymbol() to "unconfigured" units. Therefore, I'd consider it a replacement for getSymbol() and not for getName().

Note that there are 2 main use cases for names and symbols: formatting and parsing. We can only set one label for formatting, but we can use aliases to be able to parse alternate symbols or names.

dispiste commented:
Hi again @jodygarnett, I've removed kelvin label, since symbol is already defined as "K" by default, and everything compiles and passes tests (in Geotools and also in Geoserver). I've used the following command:
mvn -Drelease clean install

Note that you'll need to sync the test_units_jsr363 branch with geotools master in order to successfully compile Geoserver.

We can further refine formatting if needed after merging the branch in master (for instance if we eventually need a custom formatter for NetCDF).

I've added a new convenience method to Units class:
```java
public static String toSymbol(Unit<?> unit)
```

because as explained in my [previous comment](https://github.com/geotools/geotools/pull/1834#issuecomment-383497140), we'll wish to get the symbol most of the time, since we can't declare/alter names with current implementation.  If we really need to declare unit names (in addition to labels), we could create a mechanism to add names to DefaultUnitParser, then we could add a formatName() method in DefaultUnitParser and use this method in Units.toName().

jodygarnett commented:
rebased and merged to master

bencaradocdavies commented:
@dispiste @jodygarnett while I concur that `build.geoserver.org` is passing, and my local build is passing, Travis CI fails for GeoServer master with a suspicious mention of `kelvin`:
```
testRapNativeGribRotatedPole(org.geoserver.wcs2_0.WCSNetCDFTest)  Time elapsed: 218 sec  <<< FAILURE!
org.junit.ComparisonFailure: expected:<[K]> but was:<[kelvin]>
	at org.junit.Assert.assertEquals(Assert.java:115)
	at org.junit.Assert.assertEquals(Assert.java:144)
	at org.geoserver.wcs2_0.WCSNetCDFTest.testRapNativeGribRotatedPole(WCSNetCDFTest.java:284)

testNetcdfRotatedPole(org.geoserver.wcs2_0.WCSNetCDFTest)  Time elapsed: 175 sec  <<< FAILURE!
org.junit.ComparisonFailure: expected:<[K]> but was:<[kelvin]>
	at org.junit.Assert.assertEquals(Assert.java:115)
	at org.junit.Assert.assertEquals(Assert.java:144)
	at org.geoserver.wcs2_0.WCSNetCDFTest.testNetcdfRotatedPole(WCSNetCDFTest.java:214)
```
I wonder if the netcdf-out plugin is needed to see this?

bencaradocdavies commented:
Note also that there is another unrelated failure in the mix. The `WCSNetCDFTest` failure is new and started after this PR was merged.

bencaradocdavies commented:
I can confirm that this test is run locally in both Maven and Eclipse and passes in both cases.

bencaradocdavies commented:
@dispiste @jodygarnett I re-ran the most recent GeoServer master build on Travis CI and it passed this time, so I guess the previous failure was just a temporary Travis CI problem: https://travis-ci.org/geoserver/geoserver/builds/370120346

dispiste commented:
Thank you @jodygarnett for supporting and petting this transition, and also
@aaime and the all the rest of the team for reviewing the changes and
ensuring everything looks good.

On 24 April 2018 at 06:51, Ben Caradoc-Davies <notifications@github.com>
wrote:

> @dispiste <https://github.com/dispiste> @jodygarnett
> <https://github.com/jodygarnett> I re-ran the most recent GeoServer
> master build on Travis CI and it passed this time, so I guess the previous
> failure was just a temporary Travis CI problem: https://travis-ci.org/
> geoserver/geoserver/builds/370120346
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/geotools/geotools/pull/1834#issuecomment-383802573>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABfd3_lv6z81Kx15nm2kK4yJ9LQGeFPVks5trq9lgaJpZM4Sxnp9>
> .
>



-- 
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
   César Martínez Izquierdo
   GIS developer
   -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -
   SCOLAB: http://www.scolab.es
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
